### ISSUE TITLE
tf.nn.conv2d exhibits unexpected FP32 numerical divergence compared to reference convolution (CPU, VALID, stride=1) on M2

### ISSUE THREAD

**Issue by MartyLinZY** [tensorflow/tensorflow#106452]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.21.0

### Custom code

Yes

### OS platform and distribution

macOS 15.7.2 (arm64)

### Python version

3.12.8

### Current behavior?

We observed a systematic numerical divergence in tf.nn.conv2d when running FP32 convolutions on CPU, compared against an equivalent reference implementation (validated via cross-framework testing).

Under the following strictly controlled conditions:

FP32 inputs and kernel weights

CPU execution

stride = 1

padding = "VALID"

no bias

correct memory layout (NHWC + HWIO)

deterministic inputs (fixed random tensors)

the output values from tf.nn.conv2d differ from the reference results with relative errors up to ~1e-3, which is significantly larger than expected for deterministic FP32 CPU convolutions.

This behavior was discovered during cross-framework fuzzing, where PyTorch’s conv2d was used as a validated reference. While output shapes match exactly, the numerical discrepancy appears consistent and reproducible.

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf
import torch
import torch.nn.functional as F

# ---- Parameters ----
N, C_in, C_out = 2, 7, 8
H, W = 9, 18
k = 5
stride = 1
padding = "VALID"

# Generate identical FP32 inputs
x = np.random.randn(N, C_in, H, W).astype(np.float32)
w = np.random.randn(C_out, C_in, k, k).astype(np.float32)

# ---- TensorFlow (NHWC / HWIO) ----
x_tf = tf.convert_to_tensor(np.transpose(x, (0, 2, 3, 1)))
w_tf = tf.convert_to_tensor(np.transpose(w, (2, 3, 1, 0)))

y_tf = tf.nn.conv2d(
    x_tf,
    w_tf,
    strides=[1, stride, stride, 1],
    padding=padding
)
y_tf_np = tf.transpose(y_tf, (0, 3, 1, 2)).numpy()

# ---- Reference (PyTorch, NCHW / OIHW) ----
x_pt = torch.from_numpy(x)
w_pt = torch.from_numpy(w)

y_ref = F.conv2d(x_pt, w_pt, stride=stride, padding=0)
y_ref_np = y_ref.detach().cpu().numpy()

# ---- Comparison ----
abs_diff = np.abs(y_tf_np - y_ref_np)
rel_diff = abs_diff / (np.abs(y_ref_np) + 1e-12)

print("Output shapes:", y_tf_np.shape, y_ref_np.shape)
print("Max abs error:", abs_diff.max())
print("Max rel error:", rel_diff.max())
```

### Relevant log output

```shell

```

**Comment by gzmkl** [comment]
Linux x86_64 (Intel(R) Xeon(R) Platinum 8180 CPU @ 2.50GHz)
TensorFlow: built from the current master branch, targeted for 2.21 release
ONEDNN enabled (TF_ENABLE_ONEDNN_OPTS=1)

Output shapes: (2, 8, 5, 14) (2, 8, 5, 14)
Max abs error: 0.0
Max rel error: 0.0

So this issue must be mac specific.

**Comment by nileshau-afk** [comment]
Hi @MartyLinZY,

Apologies for the delay, and thanks for your patience. We observe the numerical divergence you're reporting for `tf.nn.conv2d` on macOS arm64. Differences in floating-point results between frameworks like TensorFlow and PyTorch, even for seemingly identical operations, are common due to variations in underlying optimized libraries (e.g., Eigen, XNNPACK for TensorFlow CPU vs. MKL-DNN, OpenBLAS for PyTorch CPU), compiler optimizations, and the exact order of operations. 

To help us investigate if this is an expected numerical variation or a specific regression, could you please test if the issue persists with the latest TensorFlow Nightly build?For more details on `tf.nn.conv2d`, you can refer to the official documentation: https://www.tensorflow.org/api_docs/python/tf/nn/conv2d. Kindly refer to this [gist](https://colab.sandbox.google.com/gist/nileshau-afk/169702e0f71165f204214ca033cbe6ce/reproduce_tf_issue_106452_nightly.ipynb).

Please let us know if the behavior changes with the nightly build.
